### PR TITLE
Added serverless capacity model to CosmosDB create

### LINKED ISSUE #1682: Support for serverless
Cosmos now has support for serverless. Can you change the create workflow to allow a customer to select serverless for throughput? Also would need to be able to detect this during database or collection creation to not show throughput options.

### PR DESCRIPTION
**Pull request by neelip**
Fixes #1682

### CODE REVIEW

File: src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    public async prompt(wizardContext: ICosmosDBWizardContext): Promise<void> {
+
+        const placeHolder: string = localize('selectDBServerMsg', 'Select a capacity model.')
+
+        wizardContext.isServerless = (await ext.ui.showQuickPick(
+            [
+                { label: localize('provisionedOption', 'Provisioned Throughput'), data: false },
+                { label: localize('serverlessOption', 'Serverless'), data: true }
+            ], { placeHolder })).data;
```
Comment (nturinski) [inline src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts:21]: It might be worth adding a "Learn More" link here.  I know the portal uses the same verbage, but provisioned throughput and serverless feel like they don't mean a whole lot to me.

The portal uses this one: https://aka.ms/cosmos-models

File: src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import { AzureWizardPromptStep } from 'vscode-azureextensionui';
+import { ext } from '../../extensionVariables';
+import { localize } from '../../utils/localize';
+import { ICosmosDBWizardContext } from './ICosmosDBWizardContext';
+
+export class CosmosDBAccountCapacityStep extends AzureWizardPromptStep<ICosmosDBWizardContext> {
+
+    public async prompt(wizardContext: ICosmosDBWizardContext): Promise<void> {
+
+        const placeHolder: string = localize('selectDBServerMsg', 'Select a capacity model.')
```
Comment (nturinski) [inline src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts:15]: Not sure if we have to go with what the portal says, but FWIW

> Choose the capacity mode for your database operations.

File: src/tree/AzureDBAPIStep.ts [hunk 3]
```diff
@@ -51,7 +52,8 @@ export class AzureDBAPIStep extends AzureWizardPromptStep<IPostgresServerWizardC
             ];
         } else {
             promptSteps = [
-                new CosmosDBAccountNameStep()
+                new CosmosDBAccountCapacityStep(),
```
Comment (nturinski) [inline src/tree/AzureDBAPIStep.ts:55]: It's odd that the capacity mode step comes before the accountNameStep.

File: src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            [
+                { label: localize('provisionedOption', 'Provisioned Throughput'), data: false },
+                { label: localize('serverlessOption', 'Serverless'), data: true }
+            ], { placeHolder })).data;
+
+    }
+
+
+    public shouldPrompt(wizardContext: ICosmosDBWizardContext): boolean {
+        return !wizardContext.isServerless;
```
Comment (nturinski) [inline src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts:27]:
```suggestion
        return wizardContext.isServerless === undefined
```

Otherwise false will be read as undefined as well.

File: src/docdb/tree/DocDBDatabaseTreeItemBase.ts [hunk 5]
```diff
@@ -96,16 +96,20 @@ export abstract class DocDBDatabaseTreeItemBase extends DocDBTreeItemBase<Contai
...
-        const minThroughput = isFixed ? minThroughputFixed : minThroughputPartitioned;
-        const throughput: number = Number(await ext.ui.showInputBox({
-            value: minThroughput.toString(),
-            ignoreFocusOut: true,
-            prompt: `Initial throughput capacity, between ${minThroughput} and ${maxThroughput}`,
-            validateInput: (input: string) => validateThroughput(isFixed, input)
-        }));
-
-        const options = { offerThroughput: throughput };
+        const options: RequestOptions = {};
```
Comment (nturinski) [inline src/docdb/tree/DocDBDatabaseTreeItemBase.ts:99]: Could you explain why this change was made?

File: src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts (hunk 1 above)
Comment (nturinski) [reply src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts:21]: You can do it like this: https://github.com/microsoft/vscode-azurestaticwebapps/blob/d09f669ba0208a66b2b56a1f20c13b4c7eab2dba/src/commands/createStaticWebApp/BuildPresetListStep.ts#L20

File: src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts (hunk 2 above)
Comment (neelip) [reply src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts:15]: I think I would prefer "Select a capacity model." as it would consistent with other placeholders. 
The "Choose the capacity mode..." might confuse the user as capacity model is selected at the server even tho it is meant for database operations. 

What if we add a "Learn More..." option to this prompt instead?

File: src/tree/AzureDBAPIStep.ts (hunk 3 above)
Comment (neelip) [reply src/tree/AzureDBAPIStep.ts:55]: accountNameStep can go before CapacityStep, it shouldn't cause a problem. 

I was going for this setup:

1. "Select an Azure Database Server."
2. and pick the next categorization which is capacity model with "Select a capacity model" prompt.
3. and then start creating the server starting with the nameStep.

File: src/docdb/tree/DocDBDatabaseTreeItemBase.ts (hunk 5 above)
Comment (neelip) [reply src/docdb/tree/DocDBDatabaseTreeItemBase.ts:99]: Options need to be defined for `throughput` when working with non-serverless database instance, and options need to be empty when working with serverless.

File: src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+export class CosmosDBAccountCapacityStep extends AzureWizardPromptStep<ICosmosDBWizardContext> {
+
+    public async prompt(wizardContext: ICosmosDBWizardContext): Promise<void> {
+
+        const placeHolder: string = localize('selectDBServerMsg', 'Select a capacity model.')
+        const picks: IAzureQuickPickItem<boolean | undefined>[] = [
+            { label: localize('provisionedOption', 'Provisioned Throughput'), data: false },
+            { label: localize('serverlessOption', 'Serverless'), data: true },
+        ];
+        const learnMore: IAzureQuickPickItem = { label: localize('learnMore', '$(link-external) Learn more...'), description: '', data: undefined };
```
Comment (nturinski) [inline src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts:21]: nit: You can just omit providing a description.

File: src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            if (pick === learnMore) {
+                await openUrl('https://aka.ms/cosmos-models');
+            }
+        } while (pick === learnMore);
+
+        if (pick.data) {
+            wizardContext.isServerless = pick.data;
+            wizardContext.telemetry.properties.isServerless = pick.data ? 'true' : 'false';
+        }
+
```
Comment (nturinski) [inline src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts:36]: nit: Lot of weird empty space here.

File: src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+
+    }
+
+
```
Comment (nturinski) [inline src/tree/CosmosDBAccountWizard/CosmosDBAccountCapacityStep.ts:41]: nit: Here as well

**Review by nturinski — APPROVED** [review]
